### ISSUE TITLE
[rllib] Enabling sample_async on remote workers when using PyTorch framework

### ISSUE THREAD

**Issue by roireshef** [ray-project/ray#10628]
### Describe your feature request

I'm thinking, given these two references below, that there's a way to enable sample_async for PyTorch as well. Since this is a major runtime booster, this feature is going to have a relatively wide impact on PyTorch/RLlib users that run long training sessions.

Thanks!

**Comment by roireshef** [comment]
- https://discuss.pytorch.org/t/is-pytorch-supposed-to-be-thread-safe/36540/2 
- https://discuss.pytorch.org/t/is-evaluating-the-network-thread-safe/37802/3

**Comment by ericl** [comment]
Thanks for reporting this. Can you provide a reproduction case with CartPole that doesn't work properly (rllib train --run=A3C --env=... etc)?

**Comment by roireshef** [comment]
Hi @ericl 

I was able to reproduce and track it to a race condition for reading and writing from/to the class variable I'm using to store the value function estimate in my TorchModelV2, let's call it `self.last_value`. The problem is that by the time a thread runs:
1. TorchModelV2.forward()
2. TorchModelV2.value_function()

another thread could override the `self.last_value` by calling TorchModelV2.forward() again. 

To reproduce, run Tune with Ray on local_mode=True (it reproduces without it, but it helps to debug in this case) with A3C. Configuration should include:
"framework": "torch",
"sample_async": True

and the validator in https://github.com/ray-project/ray/blob/master/rllib/agents/a3c/a3c.py#L60 should be commented out, since it overrides the sample_async, and I was already able to remove it and run A3C in Ray v0.7.3 with great success on sample_async mode with pytorch. I had a discussion with @sven1977 about removing this validator since this limitation is not really necessary.

Now, when you do that, end your TorchModelV2 model's forward method with something like:
```
        ...
        last_value = self.critic_net(shared_embeddings).squeeze(-1)
        self.last_value = last_value

        return self.actor_net(shared_embeddings), state
```

If you set a conditional breakpoint on the return clause that only breaks whenever the number of samples in the input is greater than 1, it's going to break only on the master thread, when the TorchModelV2.foward() method is called from the A3C loss method. This is where, if you compared `last_value` (which is only defined in the scope of that specific forward() call) is actually already different from `self.last_value` (class scope, can be overiden from outside of that specific forward() call) although one line up we assigned `last_value` to `self.last_value`.

This is an issue since the code migration to TorchModelV2, which forces the developer to use a class variable outside the scope of the forward() call. With TorchModel (original version) the forward was "atomic" in that sense that it returned the value estimates with the rest of the outputs and this was avoided.

**Comment by roireshef** [comment]
@ericl , @sven1977 - I tried to look for a use case where value_function() is called independent of the forward() call, but it seems like they are always called together, one after the other. I assume you were trying to save computation by decoupling them...? But the fact is that same design that decouples between the two really limits the performance of PyTorch + A3C. 

**Is it worth recoupling them together into forward()?**

**Comment by sven1977** [comment]
@roireshef Here is a PR that may fix this issue: https://github.com/ray-project/ray/pull/11935
Waiting for tests to pass.

**Comment by roireshef** [comment]
Hi @ericl - regarding proposal (1) in https://github.com/ray-project/ray/pull/11935#issuecomment-735016640 - is there any expectation to when this will be implemented? 

Alternatively, I could try to apply @sven1977's previous proposal to my local code in the meantime, but I'm facing a problem in which _master_ can't work well with latest nightly build in a setup like this https://docs.ray.io/en/master/development.html#building-ray-python-only - how can I tell what commit was the nightly wheel built out of so I can revert later commits on master?

**Comment by roireshef** [comment]
Hi @ericl @sven1977 - I have implemented a solution based on (1) from https://github.com/ray-project/ray/pull/11935#issuecomment-730644787

I don't think it uncovers the full potential of the async sampler runtime improvements, but I do see a 20%-50% speedup when switching from sample_async=False to sample_async=True with this feature enabled (depending on how many envs_per_worker I use).

Could you please review?
